compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -356,6 +361,31 @@ simplifyExpr world = fst . cata go
       -- e    ==>    e
       e -> (embed (fmap fst e), infoStep world (fmap snd e))
 
+exprRefs :: Expr -> Set.Set (Qualified ExprValName)
+exprRefs = cata $ \case
+    EValF x -> Set.singleton x
+    e -> fold e
+
+topoSortDefValues :: Module -> [DefValue]
+topoSortDefValues m =
+    let isLocal Qualified{..} = do
+            PRSelf <- pure qualPackage
+            guard (moduleName m == qualModule)
+            Just qualObject
+        dvalDeps = mapMaybe isLocal . Set.toList . exprRefs . dvalBody
+        dvalName = fst . dvalBinder
+        dvalNode dval = (dval, dvalName dval, dvalDeps dval)
+        sccs = G.stronglyConnComp . map dvalNode . NM.toList $ moduleValues m
+    in concatMap toList sccs
+
+simplifyDefValue :: World -> DefValue -> DefValue
+simplifyDefValue world dval = dval { dvalBody = simplifyExpr world (dvalBody dval) }
+
 simplifyModule :: World -> Module -> Module
 simplifyModule world m =
-    over moduleExpr (simplifyExpr (extendWorldSelf m world)) m
+    let step !accum dval =
+            let m' = m { moduleValues = accum }
+                w' = extendWorldSelf m' world
+                d' = simplifyDefValue w' dval
+            in NM.insert d' accum
+    in m { moduleValues = foldl' step NM.empty (topoSortDefValues m) }
```

Comment:
This looks like you are reordering the values. I guess this doesn’t matter for anything?

Reply:
I _think_ it's ok to reorder.